export_logic.go

```diff
@@ -0,0 +1,366 @@
+package libsquash
+
+import (
+	"archive/tar"
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strings"
+)
+
+var uuidRegex = regexp.MustCompile("^[a-f0-9]{64}$")
+
+func (e *export) IngestImageMetadata(instream io.Reader) error {
+	var tarReader = tar.NewReader(instream)
+	for {
+		header, err := tarReader.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		if header.Name == "." || header.Name == ".." || header.Name == "./" {
+			continue
+		}
+
+		nameParts := strings.Split(header.Name, string(os.PathSeparator))
+		if len(nameParts) == 0 {
+			continue
+		}
+
+		if len(nameParts) == 1 {
+			if nameParts[0] == "repositories" {
+				if err = json.NewDecoder(tarReader).Decode(&e.Repositories); err != nil {
+					return err
+				}
+			}
+
+			// Export may have multiple branches with the same parent.
+			// We can't handle that currently so abort.
+			for _, v := range e.Repositories {
+				commits := map[string]string{}
+				for tag, commit := range *v {
+					commits[commit] = tag
+				}
+				if len(commits) > 1 {
+					return errorMultipleBranchesSameParent
+				}
+			}
+
+			continue
+		}
+
+		uuidPart := nameParts[0]
+		fileName := nameParts[1]
+		if e.Layers[uuidPart] == nil {
+			e.Layers[uuidPart] = &layer{}
+		}
+
+		switch fileName {
+		case "json":
+			e.Layers[uuidPart].JSONHeader = header
+			if err = json.NewDecoder(tarReader).Decode(&e.Layers[uuidPart].LayerConfig); err != nil {
+				return err
+			}
+		case "layer.tar":
+			layerReader := tar.NewReader(tarReader)
+			for {
+				fileHeader, err := layerReader.Next()
+				if err != nil {
+					if err == io.EOF {
+						break
+					}
+					return err
+				}
+				filePath := nameWithoutWhiteoutPrefix(fileHeader.Name)
+				if e.fileToLayers[filePath] == nil {
+					e.fileToLayers[filePath] = []fileLoc{}
+				}
+				foundWhiteout := isWhiteout(fileHeader.Name)
+				e.fileToLayers[filePath] = append(e.fileToLayers[filePath], fileLoc{
+					uuid:     uuidPart,
+					whiteout: foundWhiteout,
+				})
+
+				if foundWhiteout {
+					e.whiteouts = append(e.whiteouts, whiteoutFile{
+						prefix: filePath,
+						uuid:   uuidPart,
+					})
+				}
+			}
+		}
+	}
+
+	e.start = e.FirstSquash()
+
+	// Can't find a previously squashed layer, default to root
+	if e.start == nil {
+		e.start = e.Root()
+	}
+
+	if e.start == nil {
+		return errorNoFROM
+	}
+
+	// TODO: optimize creation of ordered list - currently n^2, can be n
+	index := 0
+	current := e.start
+	orderMap := map[string]int{}
+	for {
+		orderMap[current.LayerConfig.ID] = index
+		index++
+		current = e.ChildOf(current.LayerConfig.ID)
+		if current == nil {
+			break
+		}
+	}
+
+	for path, fileLocs := range e.fileToLayers {
+		if len(fileLocs) > 0 {
+			greatest := fileLocs[0]
+			for _, loc := range fileLocs {
+				if orderMap[loc.uuid] > orderMap[greatest.uuid] {
```

Comment:
Please double check the logic in the refactor - this part was somewhat confusing even when I wrote it